// @strapi/content-manager/dist/admin/pages/EditView/components/FormInputs/Relations/RelationModal.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var Icons = require('@strapi/icons');
var reactIntl = require('react-intl');
var reactRouterDom = require('react-router-dom');
var styledComponents = require('styled-components');
var collections = require('../../../../../constants/collections.js');
var plugin = require('../../../../../constants/plugin.js');
var DocumentContext = require('../../../../../features/DocumentContext.js');
var DocumentRBAC = require('../../../../../features/DocumentRBAC.js');
var useDocumentLayout = require('../../../../../hooks/useDocumentLayout.js');
var documents = require('../../../../../services/documents.js');
var validation = require('../../../../../utils/validation.js');
var DocumentActions = require('../../DocumentActions.js');
var DocumentStatus = require('../../DocumentStatus.js');
var FormLayout = require('../../FormLayout.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

function getCollectionType(url) {
    const regex = new RegExp(`(${collections.COLLECTION_TYPES}|${collections.SINGLE_TYPES})`);
    const match = url.match(regex);
    return match ? match[1] : undefined;
}
const CustomModalContent = styledComponents.styled(designSystem.Modal.Content)`
  width: 90%;
  max-width: 100%;
  height: 90%;
  max-height: 100%;
`;
const [RelationModalProvider, useRelationModal] = strapiAdmin.createContext('RelationModal', {
    parentModified: false,
    depth: 0
});
const RelationModalForm = ({ relation, triggerButtonLabel })=>{
    const navigate = reactRouterDom.useNavigate();
    const { pathname, search } = reactRouterDom.useLocation();
    const { formatMessage } = reactIntl.useIntl();
    const [triggerRefetchDocument] = documents.useLazyGetDocumentQuery();
    const currentDocument = DocumentContext.useDocumentContext('RelationModalForm', (state)=>state.document);
    const rootDocumentMeta = DocumentContext.useDocumentContext('RelationModalForm', (state)=>state.rootDocumentMeta);
    const currentDocumentMeta = DocumentContext.useDocumentContext('RelationModalForm', (state)=>state.meta);
    const changeDocument = DocumentContext.useDocumentContext('RelationModalForm', (state)=>state.changeDocument);
    const documentHistory = DocumentContext.useDocumentContext('RelationModalForm', (state)=>state.documentHistory);
    const setDocumentHistory = DocumentContext.useDocumentContext('RelationModalForm', (state)=>state.setDocumentHistory);
    const [isConfirmationOpen, setIsConfirmationOpen] = React__namespace.useState(false);
    const [actionPosition, setActionPosition] = React__namespace.useState('cancel');
    const [isModalOpen, setIsModalOpen] = React__namespace.useState(false);
    // NOTE: Not sure about this relation modal context, maybe we should move this to DocumentContext?
    // Get parent modal context if it exists
    const parentContext = useRelationModal('RelationModalForm', (state)=>state);
    // Get depth of nested modals
    const depth = parentContext ? parentContext.depth + 1 : 0;
    // Check if this is a nested modal
    const isNested = depth > 0;
    const addDocumentToHistory = (document)=>setDocumentHistory((prev)=>[
                ...prev,
                document
            ]);
    const getPreviousDocument = ()=>{
        if (documentHistory.length === 0) return undefined;
        const lastDocument = documentHistory[documentHistory.length - 1];
        return lastDocument;
    };
    const removeLastDocumentFromHistory = ()=>{
        setDocumentHistory((prev)=>[
                ...prev
            ].slice(0, prev.length - 1));
    };
    const handleToggleModal = ()=>{
        if (isModalOpen) {
            setIsModalOpen(false);
            const document = {
                collectionType: rootDocumentMeta.collectionType,
                model: rootDocumentMeta.model,
                documentId: rootDocumentMeta.documentId
            };
            // Change back to the root document
            changeDocument(document);
            // Reset the document history
            setDocumentHistory([]);
            // Reset action position
            setActionPosition('cancel');
            // Read from cache or refetch root document
            triggerRefetchDocument(document, // Favor the cache
            true);
        } else {
            changeDocument(relation);
            setIsModalOpen(true);
        }
    };
    const getFullPageLink = ()=>{
        const isSingleType = currentDocumentMeta.collectionType === collections.SINGLE_TYPES;
        const queryParams = currentDocumentMeta.params?.locale ? `?plugins[i18n][locale]=${currentDocumentMeta.params.locale}` : '';
        return `/content-manager/${currentDocumentMeta.collectionType}/${currentDocumentMeta.model}${isSingleType ? '' : '/' + currentDocumentMeta.documentId}${queryParams}`;
    };
    const handleRedirection = ()=>{
        const editViewUrl = `${pathname}${search}`;
        const isRootDocumentUrl = editViewUrl.includes(getFullPageLink());
        if (isRootDocumentUrl) {
            handleToggleModal();
        } else {
            navigate(getFullPageLink());
        }
    };
    const handleConfirm = ()=>{
        if (actionPosition === 'navigate') {
            handleRedirection();
        } else if (actionPosition === 'back') {
            const previousRelation = getPreviousDocument();
            if (previousRelation) {
                removeLastDocumentFromHistory();
                changeDocument(previousRelation);
            }
        } else {
            // Add current relation to history before opening a new one in case we are opening a new one
            if (currentDocumentMeta && Object.keys(currentDocumentMeta).length > 0) {
                addDocumentToHistory(currentDocumentMeta);
            }
            handleToggleModal();
        }
    };
    return /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Form, {
        method: "PUT",
        initialValues: currentDocument.getInitialFormValues(),
        validate: (values, options)=>{
            const yupSchema = validation.createYupSchema(currentDocument.schema?.attributes, currentDocument.components, {
                status: currentDocument.document?.status,
                ...options
            });
            return yupSchema.validate(values, {
                abortEarly: false
            });
        },
        children: ({ modified, isSubmitting, resetForm })=>{
            // We don't count the root document, so history starts after 1
            const hasHistory = documentHistory.length > 1;
            return /*#__PURE__*/ jsxRuntime.jsxs(RelationModalProvider, {
                parentModified: modified,
                depth: depth,
                children: [
                    /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Modal.Root, {
                        open: isModalOpen,
                        onOpenChange: ()=>{
                            if (isModalOpen) {
                                if (modified && !isSubmitting) {
                                    setIsConfirmationOpen(true);
                                } else {
                                    handleToggleModal();
                                }
                            }
                        },
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Trigger, {
                                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Tooltip, {
                                    description: triggerButtonLabel,
                                    children: /*#__PURE__*/ jsxRuntime.jsx(CustomTextButton, {
                                        onClick: ()=>{
                                            // Check if parent modal has unsaved changes
                                            if (isNested && parentContext.parentModified) {
                                                setIsConfirmationOpen(true);
                                                // Return early to avoid opening the modal
                                                return;
                                            } else {
                                                if (modified && !isSubmitting) {
                                                    setIsConfirmationOpen(true);
                                                } else {
                                                    // Add current relation to history before opening a new one
                                                    if (currentDocumentMeta && Object.keys(currentDocumentMeta).length > 0) {
                                                        addDocumentToHistory(currentDocumentMeta);
                                                    }
                                                    handleToggleModal();
                                                }
                                                if (!isModalOpen) {
                                                    setIsModalOpen(true);
                                                }
                                            }
                                        },
                                        width: "100%",
                                        children: triggerButtonLabel
                                    })
                                })
                            }),
                            /*#__PURE__*/ jsxRuntime.jsxs(CustomModalContent, {
                                children: [
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Header, {
                                        gap: 2,
                                        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                                            justifyContent: "space-between",
                                            alignItems: "center",
                                            width: "100%",
                                            children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                                gap: 2,
                                                children: [
                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                                                        withTooltip: false,
                                                        label: "Back",
                                                        variant: "ghost",
                                                        disabled: !hasHistory,
                                                        onClick: ()=>{
                                                            setActionPosition('back');
                                                            if (modified && !isSubmitting) {
                                                                setIsConfirmationOpen(true);
                                                            } else {
                                                                const previousRelation = getPreviousDocument();
                                                                if (previousRelation) {
                                                                    removeLastDocumentFromHistory();
                                                                    changeDocument(previousRelation);
                                                                }
                                                            }
                                                        },
                                                        marginRight: 1,
                                                        children: /*#__PURE__*/ jsxRuntime.jsx(Icons.ArrowLeft, {})
                                                    }),
                                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                                        tag: "span",
                                                        fontWeight: 600,
                                                        children: formatMessage({
                                                            id: 'content-manager.components.RelationInputModal.modal-title',
                                                            defaultMessage: 'Edit a relation'
                                                        })
                                                    })
                                                ]
                                            })
                                        })
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(RelationModalBody, {
                                        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.IconButton, {
                                            onClick: ()=>{
                                                setActionPosition('navigate');
                                                if (modified && !isSubmitting) {
                                                    setIsConfirmationOpen(true);
                                                } else {
                                                    navigate(getFullPageLink());
                                                }
                                            },
                                            variant: "tertiary",
                                            label: formatMessage({
                                                id: 'content-manager.components.RelationInputModal.button-fullpage',
                                                defaultMessage: 'Go to entry'
                                            }),
                                            children: /*#__PURE__*/ jsxRuntime.jsx(Icons.ArrowsOut, {})
                                        })
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Footer, {
                                        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                                            onClick: ()=>{
                                                if (modified && !isSubmitting) {
                                                    setIsConfirmationOpen(true);
                                                } else {
                                                    handleToggleModal();
                                                }
                                            },
                                            variant: "tertiary",
                                            children: formatMessage({
                                                id: 'app.components.Button.cancel',
                                                defaultMessage: 'Cancel'
                                            })
                                        })
                                    })
                                ]
                            })
                        ]
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Dialog.Root, {
                        open: isConfirmationOpen,
                        onOpenChange: setIsConfirmationOpen,
                        children: /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.ConfirmDialog, {
                            onConfirm: ()=>{
                                handleConfirm();
                                setIsConfirmationOpen(false);
                                resetForm();
                            },
                            onCancel: ()=>{
                                setIsConfirmationOpen(false);
                            },
                            variant: "danger",
                            children: formatMessage({
                                id: 'content-manager.components.RelationInputModal.confirmation-message',
                                defaultMessage: 'Some changes were not saved. Are you sure you want to close this relation? All changes that were not saved will be lost.'
                            })
                        })
                    })
                ]
            });
        }
    });
};
const CustomTextButton = styledComponents.styled(designSystem.TextButton)`
  & > span {
    font-size: ${({ theme })=>theme.fontSizes[2]};
    width: inherit;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  }
`;
const RelationModalBody = ({ children })=>{
    const { formatMessage } = reactIntl.useIntl();
    const documentMeta = DocumentContext.useDocumentContext('RelationModalBody', (state)=>state.meta);
    const documentResponse = DocumentContext.useDocumentContext('RelationModalBody', (state)=>state.document);
    const onPreview = DocumentContext.useDocumentContext('RelationModalBody', (state)=>state.onPreview);
    const documentLayoutResponse = useDocumentLayout.useDocumentLayout(documentMeta.model);
    const plugins = strapiAdmin.useStrapiApp('RelationModalBody', (state)=>state.plugins);
    const initialValues = documentResponse.getInitialFormValues();
    const { permissions = [], isLoading: isLoadingPermissions, error } = strapiAdmin.useRBAC(plugin.PERMISSIONS.map((action)=>({
            action,
            subject: documentMeta.model
        })));
    const isLoading = isLoadingPermissions || documentLayoutResponse.isLoading || documentResponse.isLoading;
    if (isLoading && !documentResponse.document?.documentId) {
        return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Loader, {
            small: true,
            children: formatMessage({
                id: 'content-manager.ListViewTable.relation-loading',
                defaultMessage: 'Relations are loading'
            })
        });
    }
    if (error || !documentMeta.model || documentLayoutResponse.error || !documentResponse.document || !documentResponse.meta || !documentResponse.schema || !initialValues) {
        return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
            alignItems: "center",
            height: "100%",
            justifyContent: "center",
            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.EmptyStateLayout, {
                icon: /*#__PURE__*/ jsxRuntime.jsx(Icons.WarningCircle, {
                    width: "16rem"
                }),
                content: formatMessage({
                    id: 'anErrorOccurred',
                    defaultMessage: 'Whoops! Something went wrong. Please, try again.'
                })
            })
        });
    }
    const documentTitle = documentResponse.getTitle(documentLayoutResponse.edit.settings.mainField);
    const hasDraftAndPublished = documentResponse.schema?.options?.draftAndPublish ?? false;
    const props = {
        activeTab: 'draft',
        collectionType: documentMeta.collectionType,
        model: documentMeta.model,
        documentId: documentMeta.documentId,
        document: documentResponse.document,
        meta: documentResponse.meta,
        onPreview,
        fromRelationModal: true,
        fromPreview: onPreview !== undefined
    };
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Modal.Body, {
        children: /*#__PURE__*/ jsxRuntime.jsxs(DocumentRBAC.DocumentRBAC, {
            permissions: permissions,
            model: documentMeta.model,
            children: [
                /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                    alignItems: "flex-start",
                    direction: "column",
                    gap: 2,
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                            width: "100%",
                            justifyContent: "space-between",
                            gap: 2,
                            children: [
                                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                    tag: "h2",
                                    variant: "alpha",
                                    children: documentTitle
                                }),
                                /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                    gap: 2,
                                    children: [
                                        children,
                                        /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.DescriptionComponentRenderer, {
                                            props: props,
                                            descriptions: plugins['content-manager'].apis.getDocumentActions('relation-modal'),
                                            children: (actions)=>{
                                                const filteredActions = actions.filter((action)=>{
                                                    return [
                                                        action.position
                                                    ].flat().includes('relation-modal');
                                                });
                                                const [primaryAction, secondaryAction] = filteredActions;
                                                if (!primaryAction && !secondaryAction) return null;
                                                // Both actions are available when draft and publish enabled
                                                if (primaryAction && secondaryAction) {
                                                    return /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
                                                        children: [
                                                            /*#__PURE__*/ jsxRuntime.jsx(DocumentActions.DocumentActionButton, {
                                                                ...secondaryAction,
                                                                variant: secondaryAction.variant || 'secondary'
                                                            }),
                                                            /*#__PURE__*/ jsxRuntime.jsx(DocumentActions.DocumentActionButton, {
                                                                ...primaryAction,
                                                                variant: primaryAction.variant || 'default'
                                                            })
                                                        ]
                                                    });
                                                }
                                                // Otherwise we just have the save action
                                                return /*#__PURE__*/ jsxRuntime.jsx(DocumentActions.DocumentActionButton, {
                                                    ...primaryAction,
                                                    variant: primaryAction.variant || 'secondary'
                                                });
                                            }
                                        })
                                    ]
                                })
                            ]
                        }),
                        hasDraftAndPublished ? /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                            children: /*#__PURE__*/ jsxRuntime.jsx(DocumentStatus.DocumentStatus, {
                                status: documentResponse.document?.status
                            })
                        }) : null
                    ]
                }),
                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                    flex: 1,
                    overflow: "auto",
                    alignItems: "stretch",
                    paddingTop: 7,
                    children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                        overflow: "auto",
                        flex: 1,
                        children: /*#__PURE__*/ jsxRuntime.jsx(FormLayout.FormLayout, {
                            layout: documentLayoutResponse.edit.layout,
                            document: documentResponse,
                            hasBackground: false
                        })
                    })
                })
            ]
        })
    });
};

exports.RelationModalForm = RelationModalForm;
exports.getCollectionType = getCollectionType;
